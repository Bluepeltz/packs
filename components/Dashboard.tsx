import { Head } from '$fresh/runtime.ts';

import { createStyle } from 'flcss';

import Avatar from './Avatar.tsx';

import type { Schema } from '$fable/src/types.ts';

import colors from '../theme.ts';

import LinkIcon from 'icons/link.tsx';
import PlusIcon from 'icons/plus.tsx';

type User = {
  id: string;
  username: string;
  // deno-lint-ignore camelcase
  display_name?: string;
  avatar?: string;
};

export default ({ user, packs }: { user: User; packs: Schema.Pack[] }) => {
  const styles = createStyle({
    wrapper: {
      display: 'flex',
      justifyContent: 'center',
      width: '100%',
      padding: '0 10vw',
      gap: '3em',
      flexWrap: 'wrap',
      margin: '5vh 0',
    },
    card: {
      cursor: 'pointer',
      width: '128px',
      minHeight: '32px',
      color: 'inherit',
      backgroundColor: colors.embed,
      textDecoration: 'none',
      padding: '24px 16px',
      borderRadius: '8px',
      transition: 'background-color .125s, box-shadow .125s, transform .125s',
      ':hover': {
        backgroundColor: 'rgb(35, 36, 40)',
        transform: 'translateY(-8px)',
        boxShadow: '0 8px 16px rgba(0,0,0,.2)',
      },
    },
    cardImage: {
      width: '128px',
      height: '128px',
      objectFit: 'cover',
      borderRadius: '8px',
    },
    cardTitle: {
      whiteSpace: 'nowrap',
      textOverflow: 'ellipsis',
      overflow: 'hidden',
    },
    cardPlaceholder: {
      extend: 'card',
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      backgroundColor: 'transparent',
      color: colors.grey,
      border: `4px dashed ${colors.grey}`,
      '> svg': {
        width: '42px',
        height: 'auto',
      },
      ':hover': {
        transform: 'translateY(-8px)',
      },
    },
  });

  return (
    <>
      <Head>
        <style>{styles.bundle}</style>
      </Head>

      <Avatar id={user?.id} avatar={user?.avatar} />

      <div class={styles.names.wrapper}>
        {packs.map((pack) => (
          <a class={styles.names.card} href={pack.manifest.id}>
            {pack.manifest.image
              ? (
                <img
                  src={pack.manifest.image}
                  class={styles.names.cardImage}
                />
              )
              : undefined}
            <div class={styles.names.cardTitle}>
              {pack.manifest.title ?? pack.manifest.id}
            </div>
          </a>
        ))}

        <a class={styles.names.cardPlaceholder} href={'/import'}>
          <LinkIcon />
        </a>
        <a class={styles.names.cardPlaceholder} href={'/new'}>
          <PlusIcon />
        </a>
      </div>
    </>
  );
};
