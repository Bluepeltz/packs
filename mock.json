{
  "id": "bluearchive",
  "title": "Blue Archive",
  "media": {
    "new": [
      {
        "format": "VIDEO_GAME",
        "popularity": 1000000,
        "externalLinks": [],
        "description": "Blue Archive is a role-playing game developed by Nexon Games, a subsidiary of Nexon. It was released in 2021 for Android and iOS, first in Japan by Yostar and worldwide later that year by Nexon. The game is free-to-play with gacha game mechanics as a means to obtain new characters.\n\nThe player and protagonist is a teacher who was summoned to the academy city of Kivotos by the president of the General Student Council, an extrajudicial committee governing the schools. Following her sudden disappearance, criminal activity rises around Kivotos, and the player is tasked by the remaining members of the council to resolve the issues and help search for the president.\n\n(Source: Wikipedia)",
        "characters": [
          {
            "role": "MAIN",
            "characterId": "hy_m"
          },
          {
            "role": "MAIN",
            "characterId": "idla"
          },
          {
            "role": "MAIN",
            "characterId": "yz-m"
          },
          {
            "role": "SUPPORTING",
            "characterId": "sqhb"
          },
          {
            "role": "SUPPORTING",
            "characterId": "xv9p"
          },
          {
            "role": "SUPPORTING",
            "characterId": "wknt"
          },
          {
            "role": "SUPPORTING",
            "characterId": "w2jf"
          },
          {
            "role": "SUPPORTING",
            "characterId": "i2xa"
          },
          {
            "role": "SUPPORTING",
            "characterId": "c8sg"
          },
          {
            "role": "SUPPORTING",
            "characterId": "oqu7"
          },
          {
            "role": "SUPPORTING",
            "characterId": "lv7c"
          },
          {
            "role": "SUPPORTING",
            "characterId": "ahzd"
          },
          {
            "role": "SUPPORTING",
            "characterId": "gy40"
          },
          {
            "role": "SUPPORTING",
            "characterId": "8ppg"
          },
          {
            "role": "SUPPORTING",
            "characterId": "enhq"
          },
          {
            "role": "SUPPORTING",
            "characterId": "blkq"
          },
          {
            "role": "SUPPORTING",
            "characterId": "i8ku"
          },
          {
            "role": "SUPPORTING",
            "characterId": "2hdl"
          },
          {
            "role": "SUPPORTING",
            "characterId": "inhn"
          },
          {
            "role": "SUPPORTING",
            "characterId": "osup"
          },
          {
            "role": "SUPPORTING",
            "characterId": "fkle"
          },
          {
            "role": "SUPPORTING",
            "characterId": "tzf4"
          },
          {
            "role": "SUPPORTING",
            "characterId": "dmff"
          },
          {
            "role": "SUPPORTING",
            "characterId": "yhxu"
          },
          {
            "role": "SUPPORTING",
            "characterId": "r2b5"
          },
          {
            "role": "SUPPORTING",
            "characterId": "g4mf"
          },
          {
            "role": "SUPPORTING",
            "characterId": "3eyw"
          }
        ],
        "id": "xqzn",
        "relations": [],
        "title": {
          "english": "Blue Archive",
          "alternative": ["BA"]
        },
        "type": "OTHER",
        "images": [
          {
            "url": "https://f005.backblazeb2.com/file/fable-packs/22TUW0Cu"
          }
        ]
      }
    ],
    "conflicts": []
  },
  "characters": {
    "new": [
      {
        "externalLinks": [],
        "name": {
          "english": "Takanashi Hoshino",
          "alternative": ["hoshino"]
        },
        "description": "Takanashi Hoshino is the former vice president and the only third-year of Abydos High School, and the current chairwoman for the Countermeasures Committee.\n\nSpeaking like a senile aunt, she enjoys playing around and napping more than work. Because of this, she usually gets scolded by the rest of the Committee. But, once the mission starts, she protects her comrades by fighting valiantly on the front lines.",
        "age": "17",
        "id": "hy_m",
        "media": [
          {
            "role": "MAIN",
            "mediaId": "xqzn"
          }
        ],
        "images": [
          {
            "url": "https://f005.backblazeb2.com/file/fable-packs/rtfbupny"
          }
        ],
        "gender": "Female"
      }
    ],
    "conflicts": []
  },
  "maintainers": ["228674702414053386", "839261971185664011"],
  "image": "https://f005.backblazeb2.com/file/fable-packs/dkihlBqc",
  "author": "Fable Community",
  "description": "A pack containing the various characters from the game Blue Archive",
  "conflicts": [
    "anilist:161049",
    "anilist:138416",
    "anilist:160589",
    "anilist:161050",
    "anilist:143177"
  ]
}
