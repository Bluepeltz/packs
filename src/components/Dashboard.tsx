import { Head } from '$fresh/runtime.ts';

import Card from './Card.tsx';
import Avatar from './Avatar.tsx';

import Manage from './Manage.tsx';

import IconLink from 'icons/link.tsx';
import IconPlus from 'icons/plus.tsx';

import type { Schema } from 'fable/src/types.ts';

import type { PageProps } from '$fresh/server.ts';

export type User = {
  id: string;
  username: string;
  // deno-lint-ignore camelcase
  display_name?: string;
  avatar?: string;
};

export interface DashboardData {
  user?: User;
  packs: Record<string, Schema.Pack>;
}

export { Schema };

export default ({ data, url, params }: PageProps<DashboardData>) => {
  const { searchParams } = url;

  const packId = params.id;

  // deno-lint-ignore no-non-null-assertion
  const user = data.user!;

  const hasNew = searchParams.has('new');
  // const hasImport = searchParams.get('import');

  if (packId || hasNew) {
    // <Manage> is a island that doesn't have access to the deno runtime
    // this is why we query the variable here instead of directly there
    const dryRun = Deno.env.get('DRY_RUN') === '1';

    return (
      <Manage
        new={hasNew}
        dryRun={dryRun}
        pack={packId ? data.packs[packId] : undefined}
        user={user?.id}
      />
    );
  }

  return (
    <>
      <Avatar id={user?.id} avatar={user?.avatar} />

      <div class={'dashboard-wrapper'}>
        {Object.values(data.packs).map((pack) => (
          <Card key={pack.manifest.id} pack={pack} />
        ))}

        <a class={'dashboard-action-button'} href={`?new`}>
          <IconPlus />
        </a>

        <a disabled class={'dashboard-action-button'} href={`?import`}>
          <IconLink />
        </a>
      </div>
    </>
  );
};
