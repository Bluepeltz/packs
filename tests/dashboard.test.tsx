// deno-lint-ignore-file no-explicit-any

import { renderToString } from 'preact-render-to-string';

import { assertSnapshot } from '$std/testing/snapshot.ts';

import { stub } from '$std/testing/mock.ts';

import Dashboard from '../src/components/Dashboard.tsx';

import mock from './mock.json' assert { type: 'json' };

Deno.test('<Login/>', async (test) => {
  const mathStub = stub(
    Math,
    'random',
    () => 0,
  );

  try {
    await assertSnapshot(
      test,
      renderToString(
        <Dashboard
          route='/'
          url={new URL('http://localhost:8080')}
          params={{}}
          data={{
            user: {
              id: 'id',
              avatar: 'avatar_hash',
              username: 'username',
            },
            packs: [
              { manifest: mock } as any,
            ],
          }}
        />,
      ),
    );
  } finally {
    mathStub.restore();
  }
});
