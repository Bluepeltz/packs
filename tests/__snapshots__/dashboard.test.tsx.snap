export const snapshot = {};

snapshot[`<Dashboard/> 1`] = `'<img data-dialog="logout" class="flcss-logo-" src="https://cdn.discordapp.com/avatars/id/avatar_hash.png"/><i data-dialog-cb="logout" data-dialog-cb-action="hide" data-dialog-cancel="logout" style="z-index:2;position:absolute;visibility:hidden;background:#2b2d31;top:0px;left:0px;width:100%;height:100%;opacity:0.8;"></i><i data-dialog-cb="logout" data-dialog-cb-action="hide" class="flcss-container-" style="z-index:3;position:absolute;visibility:hidden;overflow:hidden auto;"><form method="post" action="/api/logout"><button class="flcss-button-" type="submit">Log Out<svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-logout" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"></path><path d="M14 8v-2a2 2 0 0 0 -2 -2h-7a2 2 0 0 0 -2 2v12a2 2 0 0 0 2 2h7a2 2 0 0 0 2 -2v-2"></path><path d="M7 12h14l-3 -3m0 6l3 -3"></path></svg></button></form></i><div class="flcss-wrapper-"><a class="flcss-placeholder-" href="?new"><svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-plus" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"></path><path d="M12 5l0 14"></path><path d="M5 12l14 0"></path></svg></a><a disabled class="flcss-placeholder-" href="?import"><svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-link" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round"><path stroke="none" d="M0 0h24v24H0z" fill="none"></path><path d="M10 14a3.5 3.5 0 0 0 5 0l4 -4a3.5 3.5 0 0 0 -5 -5l-.5 .5"></path><path d="M14 10a3.5 3.5 0 0 0 -5 0l-4 4a3.5 3.5 0 0 0 5 5l.5 -.5"></path></svg></a></div>'`;

snapshot[`<Card/> 1`] = `'<a class="flcss-card-" href="utawarerumono"><img src="https://media.tenor.com/n9YTIWj9H-gAAAAC/utawarerumono-kuon.gif" class="flcss-image-"/><div class="flcss-title-">Utawarerumono Pack</div></a>'`;
